### PR TITLE
Ampers: Mary, Alex

### PR DESCRIPTION
**Pull request by brownav**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

Feature | Feedback
--- | ---
What are the advantages of using git when collaboratively working on one code base? | With git you can push and pull on the same code to maintain version control, and work on your own machine while driving. 
Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? | Because self.highest_score_from uses the Scoring self.class method within it, also the Player class use Scoring self.class as well. It is used across a few classes in the program.
Describe an example of a test you wrote for a _nominal case_. | To return nil if a word is longer than 7 letters test is pretty straightforward, we fed in an 8 letter word --> Scrabble::Scoring.score('abcdefgh').must_be_nil 
Describe an example of a test you wrote for an _edge case_. | To take in a word for method play, we expect nil if an integer is passed in --> @john.play(20).must_be_nil
Does your code use any methods from the `Enumerable` mixin? If so, where and why was it helpful? | Our highest word score method used the map enumerable on an array of played words to return a new array with the scores of those words. It was helpful in keeping the code compact. We used the resulting array of scores to return the max score.
What was one method you and your pair used to debug code? | We used binding.pry to double-check suspicious values or NilClass errors. 
Is there a specific piece of code you'd like feedback on? | Scrabble::Scoring self.highest_score_from method has many returns and might be the most unclear at first glance. 
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | We thought our use of good coding language, continual communication, and willingness to experiment with code was accepted graciously and a great benefit to our team.

### CODE REVIEW

File: lib/scoring.rb [hunk 1]
```diff
@@ -1,9 +1,61 @@
+require 'awesome_print'
+require 'pry'
+require_relative 'player'
+
 module Scrabble
   class Scoring
+    attr_reader :word, :letter
+
+    @@letters_hash = { ["a", "e", "i", "o", "u", "l", "n", "r", "s", "t"] => 1, ["d", "g"] => 2, ["b", "c", "m", "p"] => 3, ["f", "h", "v", "w", "y"] => 4, ["k"] => 5, ["j", "x"] => 8, ["q", "z"] => 10 }
```
Comment (CheezItMan) [inline lib/scoring.rb:9]: Good data structure for this, but it probably doesn't need to be a class variable.  A constant would be a better choice.

File: lib/scoring.rb [hunk 2]
```diff
@@ -1,9 +1,61 @@
+require 'awesome_print'
+require 'pry'
+require_relative 'player'
+
 module Scrabble
   class Scoring
+    attr_reader :word, :letter
```
Comment (CheezItMan) [inline lib/scoring.rb:7]: Why have an `attr_reader` here?  You will **never** create an instance of `Scoring`  It only has class methods!

File: lib/scoring.rb [hunk 3]
```diff
@@ -1,9 +1,61 @@
...
+      word_array.each do |i|
+        @@letters_hash.keys.each do |keys|
+          if keys.include? i
+            @letter = @@letters_hash[keys]
+          end
+        end
+        score << @letter
+      end
+
+      return score.length == 7 ? score.sum + 50 : score.sum
```
Comment (CheezItMan) [inline lib/scoring.rb:33]: Good use of a ternary

File: specs/scoring_spec.rb [hunk 4]
```diff
@@ -41,22 +41,28 @@
...
-    it 'returns the only word in a length-1 array' do
-    end
+   it 'returns the only word in a length-1 array' do
+     Scrabble::Scoring.highest_score_from(["snail"]).must_equal "snail"
+   end
 
-    it 'returns the highest word if there are two words' do
-    end
+   it 'returns the highest word if there are two words' do
+     Scrabble::Scoring.highest_score_from(["dog", "quail"]).must_equal "quail"
```
Comment (CheezItMan) [inline specs/scoring_spec.rb:53]: You should also vary the order here to make sure.
```ruby
     Scrabble::Scoring.highest_score_from(["quail", "dog"]).must_equal "quail"
```

File: specs/scoring_spec.rb [hunk 5]
```diff
@@ -41,22 +41,28 @@
...
+   end
 
     it 'if tied, prefer a word with 7 letters' do
+      Scrabble::Scoring.highest_score_from(["aaaaaaa", "ggge"]).must_equal "aaaaaaa"
```
Comment (CheezItMan) [inline specs/scoring_spec.rb]: ditto

File: specs/scoring_spec.rb [hunk 6]
```diff
@@ -41,22 +41,28 @@
...
     end
 
     it 'if tied and no word has 7 letters, prefers the word with fewer letters' do
+      Scrabble::Scoring.highest_score_from(["aaaaaa", "ggg"]).must_equal "aaaaaa"
```
Comment (CheezItMan) [inline specs/scoring_spec.rb]: Read the description here.  "Prefer the word with **fewer** letters"  Huh.

Should be:
```ruby
      Scrabble::Scoring.highest_score_from(["aaaaaa", "ggg"]).must_equal "ggg"
      Scrabble::Scoring.highest_score_from(["ggg", "aaaaaa"]).must_equal "ggg"
```

File: specs/scoring_spec.rb [hunk 7]
```diff
@@ -41,22 +41,28 @@
...
     end
 
     it 'returns the first word of a tie with same letter count' do
+      Scrabble::Scoring.highest_score_from(["cat", "dog"]).must_equal "cat"
```
Comment (CheezItMan) [inline specs/scoring_spec.rb]: Also test when the order is reversed.

File: lib/player.rb [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+    def initialize(name)
+      @player_tiles = []
+      @name = name
+      @played_words = []
+      @total_score = 0
+    end
+
+    def draw_tiles(tile_bag)
+      tile_bag = Scrabble::TileBag.new
```
Comment (CheezItMan) [inline lib/player.rb]: You're getting a `TileBag` instance passed as a parameter you are **NOT** supposed to create a new instance here.  **USE** the parameter!

File: lib/player.rb [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+    def tiles
+      return @player_tiles
+    end
+
+    def play(word)
+      if @total_score >= 100
+        return false
+      else
+        @played_words << word
```
Comment (CheezItMan) [inline lib/player.rb]: You should make sure the word can be scored before adding it to the list of played words.  What if the player tries to play `!@#!@`?

File: lib/player.rb [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+        return Scrabble::Scoring.score(word)
+      end
+    end
+
+    def total_score
+      scores = @played_words.map { |i| Scrabble::Scoring.score(i) }
```
Comment (CheezItMan) [inline lib/player.rb]: Good use of `.map` and `.reduce` here!  Why have an instance variable `@total_score` then?

File: lib/player.rb [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+      total_score = scores.reduce(:+)
+      return total_score
+    end
+
+    def won?
+      if @total_score >= 100
```
Comment (CheezItMan) [inline lib/player.rb]: You have a `total_score` method that does **not** involve the instance variable.  How can this work?

File: lib/player.rb [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+        return false
+      end
+    end
+
+    def highest_scoring_word
+      return Scrabble::Scoring.highest_score_from(@played_words)
+    end
+
+    def highest_word_score
+      scores = @played_words.map { |i| Scrabble::Scoring.score(i) }
```
Comment (CheezItMan) [inline lib/player.rb]: You are missing a chance to leverage the `highest_scoring_word` method.

How could use use that to simplify this method?

File: lib/player.rb [hunk 13]
```diff
@@ -0,0 +1,61 @@
+require 'awesome_print'
+require 'pry'
+require_relative 'scoring'
+
+module Scrabble
+  class Player < Scoring
+
+    attr_reader :name, :played_words, :player_tiles
+    attr_writer :total_score
```
Comment (CheezItMan) [inline lib/player.rb]: This should **NOT** be a writer.  You're trying to hack the class to make it easier to test, but you've just made it easier to abuse.

File: specs/player_spec.rb [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+      @john.play(20).must_be_nil
+    end
+
+    it 'word is added to play array' do
+      @john.play("snail")
+      @john.played_words.must_include "snail"
+    end
+
+    it 'returns false if game is over (player has already won)' do
+      @john.total_score = 100
```
Comment (CheezItMan) [inline specs/player_spec.rb]: You shouldn't be able to directly "hack" john's total score.  Instead you should have to make him play until he wins and then try to play again.  

total_score should either be just a method or an instance variable with only an `attr_reader`

File: specs/player_spec.rb [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+
+    it 'returns score of word' do
+      @john.play("snail").must_equal 5
+      @john.play("snail").must_be_kind_of Integer
+    end
+  end
+
+  describe 'total_score' do
+    it 'returns sum of played words' do
+      @john.played_words.push("board", "cup")
```
Comment (CheezItMan) [inline specs/player_spec.rb:56]: Again you're hacking the class.  Go through the public interface and make sure that playing the words affects the list and the sum.

File: specs/player_spec.rb [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+      @john.total_score.must_equal 15
+    end
+  end
+
+  describe "won?" do
+    it "returns true if player has 100 points" do
+      @john.total_score = 100
```
Comment (CheezItMan) [inline specs/player_spec.rb]: See above

File: specs/player_spec.rb [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+  describe "highest_word_score" do
+    it 'returns the highest score played' do
+      @john.played_words.push("cat", "zebra")
+      @john.highest_word_score.must_equal 16
+    end
+  end
+
+  describe 'draw_tiles' do
+    it 'draws tiles from the tile bag' do
+      @john.draw_tiles("tiles").must_be_kind_of Array
```
Comment (CheezItMan) [inline specs/player_spec.rb]: You should be passing an instance of `TileBag` into the `Player` here.

```ruby
@john.draw_tiles(TileBag.new).must_be_kind_of Array
```

File: specs/player_spec.rb [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+      @john.draw_tiles("tiles").length.must_be :<, 8
```
Comment (CheezItMan) [inline specs/player_spec.rb]: Also after john's drawn tiles he should **always** have 7.

File: lib/tilebag.rb [hunk 19]
```diff
@@ -0,0 +1,31 @@
+require 'awesome_print'
+require 'pry'
+require_relative 'player'
+
+module Scrabble
+  class TileBag
+    attr_accessor :tilebag
+
+    def initialize
+      @tilebag = { "a" => 9, "b" => 2, "c" => 2, "d" => 4, "e" => 12, "f" => 2, "g" => 3, "h" => 2, "i" => 9, "j" => 1, "k" => 1, "l" => 4, "m" => 2, "n" => 6, "o" => 8, "p" => 2, "q" => 1, "r" => 6, "s" => 4, "t" => 6, "u" => 4, "v" => 2, "w" => 2, "x" => 1, "y" => 2, "z" => 1 }
```
Comment (CheezItMan) [inline lib/tilebag.rb]: You're giving each letter an equal chance of being drawn with this data structure, but there are **12** "E" tiles and only **1** "J".

File: lib/tilebag.rb [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+    end
+
+    def draw_tiles(num)
+      hand = []
+      num.times do
+        letter = @tilebag.keys.sample
+        @tilebag[letter] = @tilebag[letter] - 1
```
Comment (CheezItMan) [inline lib/tilebag.rb]: This could continually draw down the number of a specific tile into the negatives.

File: lib/tilebag.rb [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+        if @tilebag[letter] < 1
+          letter = @tilebag.keys.sample
```
Comment (CheezItMan) [inline lib/tilebag.rb]: This doesn't guarantee that a letter out of tiles won't be drawn.  You need to rethink your data structure here.

File: specs/tilebag_spec.rb [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+
+  before do
+    @bag = Scrabble::TileBag.new
+  end
+
+  describe 'initialize' do
+    it 'has a collection of all tiles' do
+      new_bag = Scrabble::TileBag.new
+      new_bag.must_be_instance_of Scrabble::TileBag
+      new_bag.tilebag.length.must_equal 26
```
Comment (CheezItMan) [inline specs/tilebag_spec.rb]: You don't need to test the internal structure of the `Tilebag` as a user you don't care that the `TileBag` has a hash attribute, in fact that shouldn't even be accessible to you.  Instead you should test the `tiles_remaining` method and the other methods listed in the requirements.

File: specs/tilebag_spec.rb [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+  end
+
+  describe 'draw_tiles' do
+    it 'returns an array of random tiles' do
+      @bag.draw_tiles(3).must_be_kind_of Array
+      @bag.draw_tiles(3)[2].must_be_kind_of String
+    end
+
+    it 'removes drawn tiles from tile bag' do
+      old_total = @bag.tilebag.values.sum
```
Comment (CheezItMan) [inline specs/tilebag_spec.rb]: You should also test that you can't draw more tiles out of the bag than exist.

**Comment by CheezItMan** [comment]
# Scrabble
## What We're Looking For

Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | Check, an 8-letter word would pretty much be an edge-case.  A nominal would be "dog"
Both Teammates contributed to the codebase |  Both team put in commits!
Regular Commits with meaningful commit messages |  Reasonable number of commits, focus on making the commit messages talk about the functionality added, not "waves".
**Scoring class** |
`score` method |  Check, nice work!
Uses appropriate data structure to store the letter score | Nice data structure here!
Appropriately handles edge cases |  Check
`highest_score_from` method |  Does **NOT** work, see my inline notes.  You also didn't vary the tests to be through enough to catch your bug.  Additionally your code did the exact **opposite** of one of the test descriptions.  That would have spotted the bug  See my inline notes.  
Appropriately handles edge cases |  **NOPE**
Test implementations for `highest_score_from`|  You've got some errors, see my inline notes. 
**Player class** |
Uses an array to keep track of words played |  Check
Uses existing code to create `highest_scoring_word` and `highest_word_score` |  You missed a chance to leverage one of your existing methods here.  See my inline notes.  
Returns `true` or `false` from `won?` method |  Check
Tests edge cases on the `play(word)` |  No you didn't test what happens if they play more than enough words to win.  You also didn't test invalid words, or words played in different order.  
**TileBag class** |
Uses a data structure to keep the set of default tiles in the bag |  Not a good data structure this time.  See my `TileBag` notes.  
`draw_tiles` method uses the parameter to draw a set of random tiles |  You can draw more of a letter than exist in the bag!
`tiles_remaining` method returns a count |  Check
Tests `draw_tiles` edge cases |  You should have tested trying to draw more tiles than exist in the bag.  
Summary | You've got a lot of problems here.  You need to be more through in your testing which would have pointed out some of your bugs.  You also needed to think more deeply about the `TileBag` data structure.  Lastly your `draw_tiles` method in `Player` misread the requirements.  You were to use the parameter **not** create a new instance.  You hit **most** of the learning goals, but you need to focus more on looking for every possibility in testing.
